### PR TITLE
[XLA/GPU] rsqrt is cheap and should be fused.

### PR DESCRIPTION
**Pull request by trentlo**
Please help to review the codes. Thanks.

The pattern is observed (at least) in BERT.

### CODE REVIEW

**Review by Lukious — COMMENTED** [review]
Looks Fine and it worked in my env.
It's more optimize tf module for GPU usage

**Comment by trentlo** [comment]
@sanjoy @thomasjoerg 
Could either of you help to take a look or suggest a reviewer? Thanks!

**Review by sanjoy — COMMENTED** [review]
Hi Trent,

Can you share some specific cases where this helps?  The last time I wanted to make this change I concluded https://github.com/tensorflow/tensorflow/commit/379268e9f4cbccfc46827408a0e67896c75af5b4 was a more principled fix instead.

[edit: To clarify, I'm implying that tuning the heuristic added in  https://github.com/tensorflow/tensorflow/commit/379268e9f4cbccfc46827408a0e67896c75af5b4 might be more effective.]

**Comment by trentlo** [comment]
The heuristic is good (I like it). However, rsqrt (or div) is just mapped to one hardware instruction unlike other instrinsics, which will be expanded into a bunch of instructions when linking in libdevice. So, I think that marking cheap instructions cheap is orthogonal to the heuristic (which better deals with real expensive instructions).

I see the case in layer norm.
image

File: tensorflow/compiler/xla/service/gpu/instruction_fusion.cc [hunk 1]
```diff
@@ -29,12 +29,23 @@ limitations under the License.
 namespace xla {
 namespace gpu {
 
+bool ElementIsF32OrF16(const Shape& shape) {
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/instruction_fusion.cc:32]: `static` or put under anonymous namespace.

File: tensorflow/compiler/xla/service/gpu/instruction_fusion.cc [hunk 2]
```diff
@@ -29,12 +29,23 @@ limitations under the License.
...
+
 /*static*/ bool GpuInstructionFusion::IsExpensive(
     const HloInstruction& instruction) {
-  // We say that floating-point division is cheap on the GPU.
-  if (instruction.opcode() == HloOpcode::kDivide &&
-      ShapeUtil::ElementIsFloating(instruction.shape())) {
-    return false;
+
+  // We say that some floating-point math ops are cheap on the GPU.
+  switch (instruction.opcode()) {
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/instruction_fusion.cc:41]: Please add the rationale you mentioned in the PR (that these lower to single instructions).

File: tensorflow/compiler/xla/service/gpu/instruction_fusion.cc (hunk 1 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/instruction_fusion.cc:32]: My oversight. Thanks for the catch. Will update it soon,

File: tensorflow/compiler/xla/service/gpu/instruction_fusion.cc (hunk 2 above)
Comment (trentlo) [reply tensorflow/compiler/xla/service/gpu/instruction_fusion.cc:41]: Make sense. Will add.

**Review by trentlo — COMMENTED** [review]
Will update the PR soon.

**Comment by trentlo** [comment]
Updated. Please help to take a look again. Thanks!

**Review by sanjoy — APPROVED** [review]

**Comment by sanjoy** [comment]
Hi @trentlo ,

This seems to regress a variant of resnet only on V100 by around 10%.  Here is the pre-optimization HLO: https://gist.github.com/sanjoy/8161733b3e8f303d2f81b38814661f9a

Can you PTAL?  Let me know if you can't reproduce the regression.

**Comment by trentlo** [comment]
> Hi @trentlo ,
> 
> This seems to regress a variant of resnet only on V100 by around 10%. Here is the pre-optimization HLO: https://gist.github.com/sanjoy/8161733b3e8f303d2f81b38814661f9a
> 
> Can you PTAL? Let me know if you can't reproduce the regression.

I'd guess that it interacts with the fusion heuristic and produces a surprising fusion result. I will take a look.

**Comment by trentlo** [comment]
> Hi @trentlo ,
> 
> This seems to regress a variant of resnet only on V100 by around 10%. Here is the pre-optimization HLO: > 
> Can you PTAL? Let me know if you can't reproduce the regression.

@sanjoy, I instead see 1% speedup with this PR (according to the perf numbers reported by xla_profile). See the attached log file for some [details](https://gist.github.com/sanjoy/8161733b3e8f303d2f81b38814661f9a).

Are you sure if the regression is related to this PR? Also, I wonder if you see any perf gain?

**Comment by sanjoy** [comment]
> Are you sure if the regression is related to this PR? Also, I wonder if you see any perf gain?

Could have been operator error, trying again.

**Pull request merged**